File: sdks/typescript/src/task.ts
// Copyright 2026 ConveyorQ
//
// SPDX-License-Identifier: Apache-2.0

import { ContentType, decodeJson, decodeText, type Payload } from "./codec.js";

/** Fields used to construct a {@link Task}. */
export interface TaskInit {
  type: string;
  contentType: string;
  data: Uint8Array;
  metadata?: Record<string, string>;
  id?: string;
  queue?: string;
  retried?: number;
  maxRetry?: number;
}

/**
 * Task is both what a producer enqueues and what a handler receives. A producer
 * builds one with {@link newTask}; a worker is handed one per dispatch. The
 * payload is opaque bytes plus a content type — decode it with {@link Task.json}
 * (or {@link Task.payload} for raw bytes).
 */
export class Task {
  /** The handler routing key, e.g. "email:welcome". */
  readonly type: string;
  /** How the payload is encoded, e.g. "application/json". */
  readonly contentType: string;
  /** User tags and trace propagation carried with the task. */
  readonly metadata: Record<string, string>;
  /** The task id; assigned by the server, set on dispatched tasks. */
  readonly id: string;
  /** The queue the task belongs to; set on dispatched tasks. */
  readonly queue: string;
  /** How many times this task has already been retried. */
  readonly retried: number;
  /** The retry budget before the task is dead-lettered. */
  readonly maxRetry: number;

  private readonly data: Uint8Array;

  constructor(init: TaskInit) {
    this.type = init.type;
    this.contentType = init.contentType;
    this.metadata = init.metadata ?? {};
    this.data = init.data;
    this.id = init.id ?? "";
    this.queue = init.queue ?? "";
    this.retried = init.retried ?? 0;
    this.maxRetry = init.maxRetry ?? 0;
  }

  /** The raw payload bytes. */
  payload(): Uint8Array {
    return this.data;
  }

  /** Decode a JSON payload into a value. Throws if the content type is not JSON. */
  json<T>(): T {
    return decodeJson<T>(this.data, this.contentType);
  }

  /** Decode the payload bytes as a UTF-8 string. */
  text(): string {
    return decodeText(this.data);
  }
}

/**
 * newTask builds a task to enqueue. Construct the payload with a codec —
 * `json(value)` (the default) or `bytes(data)`:
 *
 * ```ts
 * client.enqueue(newTask("email:welcome", json({ userId: 42 })));
 * ```
 */
export function newTask(type: string, payload: Payload, options?: { metadata?: Record<string, string> }): Task {
  const init: TaskInit = { type, contentType: payload.contentType || ContentType.JSON, data: payload.data };
  if (options?.metadata !== undefined) {
    init.metadata = options.metadata;
  }

  return new Task(init);
}

File: sdks/go/task.go
// Copyright 2026 ConveyorQ
//
// SPDX-License-Identifier: Apache-2.0

package conveyor

import (
	"encoding/json"
	"fmt"

	"google.golang.org/protobuf/proto"
)

// Task is a unit of work: what clients enqueue and what handlers receive.
// On the enqueue side only the type and payload are set; the remaining
// accessors are filled in for handler-side tasks.
type Task struct {
	// id is the task ULID; empty until the server assigns one.
	id string
	// queue is the queue the task belongs to.
	queue string
	// taskType is the handler routing key, e.g. "email:welcome".
	taskType string
	// payload is the encoded payload.
	payload []byte
	// contentType describes the payload encoding.
	contentType string
	// metadata carries user tags and trace propagation.
	metadata map[string]string
	// retried is how many times the task has been retried.
	retried int
	// maxRetry is the task's retry budget.
	maxRetry int
	// payloadErr is a deferred payload encoding failure.
	payloadErr error
}

// NewTask builds an enqueueable task of the given type. A payload encoding
// failure is carried inside and surfaces from Client.Enqueue.
func NewTask(taskType string, payload Payload) *Task {
	return &Task{
		taskType:    taskType,
		payload:     payload.data,
		contentType: payload.contentType,
		payloadErr:  payload.err,
	}
}

// ID returns the task ULID; empty before the task is enqueued.
func (t *Task) ID() string { return t.id }

// Type returns the handler routing key.
func (t *Task) Type() string { return t.taskType }

// Queue returns the queue the task belongs to; empty before enqueue.
func (t *Task) Queue() string { return t.queue }

// Payload returns the raw encoded payload.
func (t *Task) Payload() []byte { return t.payload }

// ContentType returns the payload encoding, e.g. "application/json".
func (t *Task) ContentType() string { return t.contentType }

// Retried returns how many times the task has been retried so far.
func (t *Task) Retried() int { return t.retried }

// MaxRetry returns the task's retry budget.
func (t *Task) MaxRetry() int { return t.maxRetry }

// Metadata returns the task's metadata tags; the map must not be mutated.
func (t *Task) Metadata() map[string]string { return t.metadata }

// Bind decodes the payload into v according to the content type: JSON
// payloads unmarshal into any value, binary payloads bind to *[]byte, and
// protobuf payloads bind to a proto.Message. Handlers typically wrap a
// Bind failure in SkipRetry: a payload that cannot decode now never will.
func (t *Task) Bind(v any) error {
	switch t.contentType {
	case ContentTypeJSON:
		if err := json.Unmarshal(t.payload, v); err != nil {
			return fmt.Errorf("conveyor: binding JSON payload: %w", err)
		}

		return nil

	case ContentTypeBytes:
		target, ok := v.(*[]byte)
		if !ok {
			return fmt.Errorf("conveyor: binary payloads bind to *[]byte, got %T", v)
		}

		*target = t.payload

		return nil

	case ContentTypeProto:
		message, ok := v.(proto.Message)
		if !ok {
			return fmt.Errorf("conveyor: protobuf payloads bind to a proto.Message, got %T", v)
		}

		if err := proto.Unmarshal(t.payload, message); err != nil {
			return fmt.Errorf("conveyor: binding protobuf payload: %w", err)
		}

		return nil

	default:
		return fmt.Errorf("conveyor: no codec for content type %q", t.contentType)
	}
}
